File: src/PhpChords/Printers/CssPrinter.php
<?php

namespace PhpChords\Printers;

class CssPrinter
{
    const BREAK_TEMPLATE = '<br/>\n';
    const SPACE_TEMPLATE = '&nbsp;';

    // Workaround to spec it!
    public function static_print(string $song_text)
    {
        return self::print($song_text);
    }

    public static function print(string $song_text)
    {
        $lines = Line::generate_all($song_text);
        $html_lines =  array_map(function ($line) {
            $html_line = array_map(function ($part_chord) use ($line) {
                list($part, $chord) = $part_chord;
                $classes = self::_clasesAttribute($chord, $line->is_chorus());
                $chords = self::_chordAttributes($chord);
                return self::_format($classes, $chords, $part);
            }, $line->to_array());
            return implode($html_line);
        }, $lines);

        return implode(self::BREAK_TEMPLATE, $html_lines);
    }

    private static function _clasesAttribute($chord, bool $chorus)
    {
        $classes = [];
        if ($chord != null)
            $classes[] = "lyric chord";
        if ($chorus)
            $classes[] = "chorus";

        return count($classes) == 0 ? '' : " class='" . trim(implode(' ', $classes)) . "'";
    }

    private static function _chordAttributes($chord)
    {
        return $chord ? " data-chord='{$chord}' data-orig-chord='{$chord}'" : '';
    }

    private static function _format($classes, $chords, $part)
    {
        return "<span{$classes}{$chords}>{$part}</span>";
    }
}

File: TO_BE_PORTED/printers/css_printer.rb
# frozen_string_literal: true

module Chordpro
  module Printers
    ##
    # Printer of a song using the new css template
    class CssPrinter < BasePrinter
      class << self
        # chorus, data- attributes, part
        TEMPLATE = '<span class="%s"%s>%s</span>'
        BREAK_TEMPLATE = '<br/>'
        SPACE_TEMPLATE = '&nbsp;'

        def print(song_text)
          Line.generate_all(song_text).each_with_index.map do |line|
            line.to_h.values.map do |value|
              text, chord = value
              format(TEMPLATE, classes(line.chorus?, chord), data_chord(chord), padded_text(text, chord))
            end.join
          end.join("#{BREAK_TEMPLATE}\n") + BREAK_TEMPLATE
        end

        private

        def padded_text(text, chord)
          if !chord.nil? && text.length < chord.length
            return text + (SPACE_TEMPLATE * ((chord.length - text.length) + 2))
          end

          text
        end

        def classes(chorus, chord)
          ret = chord ? 'lyric' : ''
          ret += ' chord' if chord
          ret += ' chorus' if chorus

          ret
        end

        def data_chord(chord)
          return '' if chord.nil?

          " data-chord='#{chord}' data-orig-chord='#{chord}'"
        end
      end
    end
  end
end
